akka-docs/src/main/paradox/typed/reliable-delivery.md

```diff
@@ -0,0 +1,358 @@
+
+### Work pulling delivery semantics
+
+For work pulling the order of the messages should not matter, because each message is routed randomly
+to one of the workers with demand.
+
+As long as neither producers nor workers crash (or workers being removed for other reasons) the messages are
+delivered to the workers without loss or duplicates. Meaning effectively once processing without any
+business level deduplication.
+
+Unconfirmed messages may be lost if the producer crashes. To avoid that you need to enable the @ref:[durable
+queue](#durable-producer) on the producer side. The stored unconfirmed messages will be redelivered when the
+corresponding producer is started again. Those messages may be routed to different workers than before
+and some of them may have already been processed but the fact that they were confirmed had not been stored
+yet. Meaning at-least once delivery.
+
+If a worker crashes or is stopped gracefully the unconfirmed messages will be redelivered to other workers.
+In that case some of these may already have been processed by the previous worker. Meaning at-least once delivery.
+
+## Sharding
+
+Reliable delivery between a producer actor sending messages to @ref:[sharded](cluster-sharding.md) consumer
+actor receiving the messages.
+
+![delivery-work-sharding-1.png](./images/delivery-sharding-1.png)
+
+and sending to another entity
+
+![delivery-work-sharding-2.png](./images/delivery-sharding-2.png)
+
+and sending from another producer (different node)
+
+![delivery-work-sharding-3.png](./images/delivery-sharding-3.png)
+
+The @apidoc[ShardingProducerController] should be used together with @apidoc[ShardingConsumerController].
+
+A producer can send messages via a `ShardingProducerController` to any `ShardingConsumerController`
+identified by an `entityId`. A single `ShardingProducerController` per `ActorSystem` (node) can be
+shared for sending to all entities of a certain entity type. No explicit registration is needed
+between the `ShardingConsumerController` and `ShardingProducerController`.
+
+The producer actor will start the flow by sending a `ShardingProducerController.Start`
+message to the `ShardingProducerController`.
+
+The `ShardingProducerController` sends `RequestNext` to the producer, which is then allowed
+to send one message to the `ShardingProducerController`. Thereafter the producer will receive a
+new `RequestNext` when it's allowed to send one more message.
+
+In the @apidoc[ShardingProducerController.RequestNext] message there is information about which entities
+that have demand. It is allowed to send to a new `entityId` that is not included in the `RequestNext.entitiesWithDemand`.
+If sending to an entity that doesn't have demand the message will be buffered. This support for buffering
+means that it is even allowed to send several messages in response to one `RequestNext` but it's recommended to
+only send one message and wait for next `RequestNext` before sending more messages.
```

Comment:
Does that mean that I can send many messages to different `entityIds` for one single `RequestNext`? What will be the impact of the confirmation in that case? 

The PC sends one request to P, P sends many messages to different entities. The entities confirm them all to the CC, then it will send many Requests back to the PC. This could cause an exponential increase in traffic, no?

Reply:
I think this breaks the flow control. If I can send more reponses then requested, then I'm sending messages to actors that didn't request it.

Reply:
For sharding we can't be too strict with the demand, because it must be possible to send to new entities. Therefore it has support for buffering, but it's up to the user to use it wisely by inspecting the information in the RequestNext.